### PR TITLE
mvcc: Limit number of bbolt read transactions opened by committed reads

### PR DESCRIPTION
**Pull request by jpbetz**

### CODE REVIEW

File: mvcc/backend/backend.go [hunk 1]
```diff
@@ -330,10 +337,20 @@ func (b *backend) run() {
...
-	for i := 0; i < 100; i++ {
+	// TODO(jpbetz): This has the potential to backlog indefinitely under heavly load.
+	// If we're going to impose a limit here. We might want to do more to ensure we're
+	// managing context deadlines and cancelations also.
+
+	// Since bbolt already keeps track of how many concurrent open read txns there are,
+	// we can use its data to enforce our limit.
+	concurrentReadTxns := b.db.Stats().OpenTxN
+
+	for i := 0; i < (maxConcurrentReadTxns - concurrentReadTxns); i++ {
```
Comment (jingyih) [inline mvcc/backend/backend.go]: We probably need to run some simple benchmarks to find out what is the typical value for concurrentReadTxns.

**Review by jingyih — COMMENTED** [review]
Looks good!  We need to make sure maxConcurrentReadTxns is larger than typical value of concurrentReadTxns.

File: mvcc/backend/backend.go (hunk 1 above)
Comment (jpbetz) [reply mvcc/backend/backend.go]: It's 1 + number of committed reads. Double checked it with a benchmark. Adding a gauge metric to this PR for the `concurrentReadTxns` value.

Comment (jpbetz) [reply mvcc/backend/backend.go]: But we will need to determine what a reasonable limit might be. I'm not 100% convinced that # of concurrent requests is sufficient, got an idea for another approach where we weight requests by their response count.

**Comment by jpbetz** [comment]
Merging this for image build. Post-commit reviews still welcome.

**Pull request merged**